GH-35686:[Go] Add AppendTime to TimestampBuilder

yevgenypats opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* Closes: #35686

zeroshade reviewed (changes requested):
Please include the updates to the tmpl files also. 

Otherwise this LGTM, though it would be good to add tests for all the unit types, not just for `Second`.

zeroshade commented:
@yevgenypats Do we want to expand the tests to test more units than just `Second`? Or do you think we're fine with just testing `Second`?

yevgenypats commented:
> @yevgenypats Do we want to expand the tests to test more units than just `Second`? Or do you think we're fine with just testing `Second`?

I can add but maybe we want to move TimestampBuilder out of the template as it becomes quite hard to develop inside a template and Timestamp starts to diverge from all other numeric types. wdyt?

zeroshade commented:
@yevgenypats Makes sense to me, feel free to go ahead and do that.

candiduslynx commented:
Hi @zeroshade!

I'm overtaking this PR from @yevgenypats.

I've extracted timestamp from numeric gen (array, builder, tests, traits).

Additionally, I propose to introduce a helper func for creation of `arrow.Timestamp` from `time.Time` and `arrow.TimeUnit`.

The logic was tested and introduced in https://github.com/cloudquery/arrow/pull/17 (in this PR: 1689eee0ec81da6e62f545eb06f8f71ed3f69a22).

yevgenypats commented:
> Hi @zeroshade!
> 
> I'm overtaking this PR from @yevgenypats.
> 
> I've extracted timestamp from numeric gen (array, builder, tests, traits).
> 
> Additionally, I propose to introduce a helper func for creation of `arrow.Timestamp` from `time.Time` and `arrow.TimeUnit`.
> 
> The logic was tested and introduced in [cloudquery#17](https://github.com/cloudquery/arrow/pull/17) (in this PR: [1689eee](https://github.com/apache/arrow/commit/1689eee0ec81da6e62f545eb06f8f71ed3f69a22)).
> 
> UPD: it seems there's a [flaky](https://github.com/apache/arrow/actions/runs/5077433733/jobs/9120706233?pr=35687#step:5:739) test (`go/arrow/flight/flightsql/driver/driver_test.go:741`, `TestPreparedStatementNoSchema`)

Thanks @candiduslynx !

candiduslynx commented on go/arrow/type_traits_test.go:
```diff
@@ -233,3 +233,37 @@ func TestMonthDayNanoIntervalTraits(t *testing.T) {
+		arrow.TimestampTraits.PutValue(b2[beg:end], arrow.Timestamp(i))
+	}
+
+	if !reflect.DeepEqual(b1, b2) {
```
I just copy-pasted from the generated code. I can fix it here, if you insist.

candiduslynx replied:
Done in e57e2ae596a072ec3e0d1da6460f7d21e86cbd55

zeroshade reviewed (approved):
LGTM Thanks!
